### PR TITLE
Add code generation for all wire types for stream encoding

### PR DESCRIPTION
**Pull request by dianale31**
Creates the code generation for all the primitive types, enums, typedefs, structs, and container types. We use the same testing utils as the reader implementation in #496 (hopefully that will reduce rebase conflicts).

### CODE REVIEW

**Comment by dianale31** [comment]
Benchmark tests for stream encoding show significant savings for bytes allocated as well as CPU time:

```
BenchmarkRoundTrip/PrimitiveOptionalStruct/Encode-8  	  337783	      2997 ns/op	     704 B/op	       1 allocs/op
BenchmarkRoundTrip/PrimitiveOptionalStruct/Stream_Encode-8         	 1000000	      1059 ns/op	       0 B/op	       0 allocs/op
BenchmarkRoundTrip/PrimitiveOptionalStruct/Decode-8                	  385268	      3520 ns/op	    1400 B/op	      14 allocs/op
BenchmarkRoundTrip/Graph/Encode-8                                  	  328562	      3665 ns/op	    1705 B/op	      11 allocs/op
BenchmarkRoundTrip/Graph/Stream_Encode-8                           	  911430	      1483 ns/op	       0 B/op	       0 allocs/op
BenchmarkRoundTrip/Graph/Decode-8                                  	  158534	      7009 ns/op	    2776 B/op	      32 allocs/op
BenchmarkRoundTrip/ContainersOfContainers/Encode-8                 	   46626	     25660 ns/op	    1305 B/op	      18 allocs/op
BenchmarkRoundTrip/ContainersOfContainers/Stream_Encode-8          	   83979	     14972 ns/op	       0 B/op	       0 allocs/op
BenchmarkRoundTrip/ContainersOfContainers/Decode-8                 	   20068	     61197 ns/op	   12315 B/op	     164 allocs/op
```

File: gen/stream.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	case *compile.StringSpec:
+		return fmt.Sprintf("%s.WriteString(%s)", sw, varName), nil
+	case *compile.BinarySpec:
+		return fmt.Sprintf("%s.WriteBinary(%s)", sw, varName), nil
+	case *compile.MapSpec:
+		encoder, err := sg.mapG.Encode(g, s)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprintf("%s(%s, %s)", encoder, varName, sw), nil
```
Comment (abhinav) [inline gen/stream.go:61]: optional: the contract for functions that return errors is that if err != nil, the other value is meaningless so you can shorten this to:
```suggestion
		return fmt.Sprintf("%s(%s, %s)", encoder, varName, sw), err
```

File: idl/internal/lex.go [hunk 2]
```diff
@@ -20,9 +20,6 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-
-
```
Comment (abhinav) [inline idl/internal/lex.go:24]: unexpected changes here? your IDE might be running gofmt on files you didn't touch.

**Review by abhinav — COMMENTED** [review]
Mostly LGTM

RE: benchmark: would you mind running a before and after and comparing them with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat?

File: gen/enum.go [hunk 3]
```diff
@@ -162,6 +163,20 @@ func enum(g Generator, spec *compile.EnumSpec) error {
 			return &<$v>
 		}
 
+		<$sw := newVar "sw">
+		// Encode encodes <$enumName> directly to the wire.
```
Comment (r-hang) [inline gen/enum.go:167]: Is to bytes more appropriate since 'the wire' could be confused for the intermediate representation to avoid?

File: gen/enum.go [hunk 4]
```diff
@@ -162,6 +163,20 @@ func enum(g Generator, spec *compile.EnumSpec) error {
...
+		//
+		//   sWriter := BinaryStreamer.Writer(writer)
+		//
+		//   var <$v> <$enumName>
+		//   if err := <$v>.Encode(sWriter); err != nil {
+		//     return err
+		//   }
```
Comment (r-hang) [inline gen/enum.go:174]: Could this just be `return <$v>.Encdoe(sWriter)`

File: gen/roundtrip_test.go [hunk 5]
```diff
@@ -81,3 +83,53 @@ func assertBinaryRoundTrip(t *testing.T, w wire.Value, message string) (wire.Val
...
+
+func testRoundTripCombos(t *testing.T, x thriftType, v wire.Value, msg string) {
+	t.Helper()
+
+	useStreaming := []struct {
+		encode bool
+		decode bool
+	}{
+		{false, false},
+		//{false, true},
```
Comment (r-hang) [inline gen/roundtrip_test.go:95]: Are these cases supposed to be commented out vs removed?

File: gen/util_for_test.go [hunk 6]
```diff
@@ -37,3 +38,13 @@ type thriftType interface {
...
 }
+
+// streamingThriftType is implemented by all generated types that know how to
+// write and read themselves to the Thrift Protocol, skipping over the
+// intermediary wire.Type
+type streamingThriftType interface {
+	thriftType
+
+	Encode(stream.Writer) error
+	//Decode(stream.Reader) error
```
Comment (r-hang) [inline gen/util_for_test.go:49]: Decode?

File: gen/roundtrip_test.go (hunk 5 above)
Comment (dianale31) [reply gen/roundtrip_test.go:95]: Commented out, this will eventually be merged with #496 which will provide the other half of the test suite.

File: gen/util_for_test.go (hunk 6 above)
Comment (dianale31) [reply gen/util_for_test.go:49]: Commented out for now until we can merge #496 and this PR together.

File: Makefile [hunk 7]
```diff
@@ -127,7 +127,7 @@ space +=
 
 .PHONY: cover
 cover:
-	go test -v -race -covermode=atomic -coverprofile cover.full.out -coverpkg=./... ./...
+	go test -timeout 100s -v -race -covermode=atomic -coverprofile cover.full.out -coverpkg=./... ./...
```
[the remark this answers is not in the document]
Comment (dianale31) [reply Makefile]: Was running into some issues with timeouts on the testing part of the build last night. But, it looks like it was a temporary thing. Removed!

**Comment by dianale31** [comment]
Benchmarks for before and after adding the new stream implementation:
```
name                                        old time/op    new time/op    delta
RoundTrip/PrimitiveOptionalStruct/Encode-8    2.03µs ± 3%    2.17µs ± 2%    +6.94%  (p=0.008 n=5+5)
RoundTrip/PrimitiveOptionalStruct/Decode-8    2.99µs ± 6%    3.33µs ± 3%   +11.58%  (p=0.008 n=5+5)
RoundTrip/PrimitiveOptionalStruct/Stream_Encode-8             1.99µs ±11%
RoundTrip/Graph/Encode-8                      3.62µs ± 7%    3.81µs ± 2%      ~     (p=0.095 n=5+5)
RoundTrip/Graph/Decode-8                      6.00µs ± 2%   11.03µs ± 2%   +83.85%  (p=0.008 n=5+5)
RoundTrip/Graph/Stream_Encode-8                              2.41µs ± 2%
RoundTrip/ContainersOfContainers/Encode-8     23.8µs ± 4%    25.7µs ± 8%    +8.11%  (p=0.008 n=5+5)
RoundTrip/ContainersOfContainers/Decode-8     53.8µs ± 2%   147.5µs ± 5%  +173.97%  (p=0.008 n=5+5)
RoundTrip/ContainersOfContainers/Stream_Encode-8             16.9µs ± 3%

name                                        old alloc/op   new alloc/op   delta
RoundTrip/PrimitiveOptionalStruct/Encode-8      704B ± 0%      704B ± 0%      ~     (all equal)
RoundTrip/PrimitiveOptionalStruct/Decode-8    1.40kB ± 0%    1.45kB ± 0%    +3.43%  (p=0.008 n=5+5)
RoundTrip/PrimitiveOptionalStruct/Stream_Encode-8            0.00B 
RoundTrip/Graph/Encode-8                      1.70kB ± 0%    1.71kB ± 0%    +0.06%  (p=0.008 n=5+5)
RoundTrip/Graph/Decode-8                      2.78kB ± 0%    3.56kB ± 0%   +28.25%  (p=0.008 n=5+5)
RoundTrip/Graph/Stream_Encode-8                              0.00B
RoundTrip/ContainersOfContainers/Encode-8     1.30kB ± 0%    1.30kB ± 0%    +0.08%  (p=0.008 n=5+5)
RoundTrip/ContainersOfContainers/Decode-8     12.3kB ± 0%    28.6kB ± 0%  +132.41%  (p=0.008 n=5+5)
RoundTrip/ContainersOfContainers/Stream_Encode-8              0.00B

name                                        old allocs/op  new allocs/op  delta
RoundTrip/PrimitiveOptionalStruct/Encode-8      1.00 ± 0%      1.00 ± 0%      ~     (all equal)
RoundTrip/PrimitiveOptionalStruct/Decode-8      14.0 ± 0%      15.0 ± 0%    +7.14%  (p=0.008 n=5+5)
RoundTrip/PrimitiveOptionalStruct/Stream_Encode-8               0.00
RoundTrip/Graph/Encode-8                        11.0 ± 0%      11.0 ± 0%      ~     (all equal)
RoundTrip/Graph/Decode-8                        32.0 ± 0%      63.0 ± 0%   +96.88%  (p=0.008 n=5+5)
RoundTrip/Graph/Stream_Encode-8                                 0.00
RoundTrip/ContainersOfContainers/Encode-8       18.0 ± 0%      18.0 ± 0%      ~     (all equal)
RoundTrip/ContainersOfContainers/Decode-8        164 ± 0%       837 ± 0%  +410.37%  (p=0.008 n=5+5)
RoundTrip/ContainersOfContainers/Stream_Encode-8                0.00
```

File: gen/roundtrip_test.go [hunk 8]
```diff
@@ -27,9 +27,11 @@ import (
 	"testing"
 
 	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/protocol/binary"
 	"go.uber.org/thriftrw/wire"
```
Comment (witriew) [inline gen/roundtrip_test.go:32]: nit: newline

File: gen/roundtrip_test.go (hunk 5 above)
Comment (witriew) [reply gen/roundtrip_test.go:95]: 👍 yep - I'll be taking care of the merging.

**Review by abhinav — APPROVED** [review]
Looks good!
